### ISSUE TITLE
Add log₂(pixel_count) as a training-time signal to reduce zensim's scale bias

### ISSUE THREAD

**Issue by lilith** [imazen/zensim#13]
## Background

The scale-invariance evaluation in #12 (results in [`docs/scale-invariance.md`](https://github.com/imazen/zensim/blob/main/docs/scale-invariance.md)) measured how zensim's score drifts when the same content is rendered at multiple resolutions and distorted identically. Two findings from that work motivate this issue:

1. **Photo codec drift**: median |β| ≈ 0.46 / octave (zensim) vs ≈ 0.69 (SSIMULACRA2) vs ≈ 0.0001 (DSSIM). Zensim is about 4% of its codec IQR per octave.
2. **Text/UI codec drift**: jumps to 0.71 / octave, mean signed +0.78 — zensim systematically under-scores low-resolution text. SSIMULACRA2's R² collapses to 0.34, suggesting its behavior on text isn't even monotonic in log₂(pixel_count).

Drift is real but modest for natural photographs. For text/UI it's large enough to matter — and the existing zensim weights have no way to know about it because the training data isn't resolution-balanced and `log₂(pixel_count)` isn't a feature.

## Proposal

Add resolution awareness to the zensim training pipeline at three layers:

### 1. Data layer

- Generate a **resolution-balanced training augmentation**: for each source, include the same content at ≥3 pyramid levels with the same codec/quality, and weight them so each resolution gets equal influence in the SROCC objective. Existing CSV is by-tile (one resolution per tile), which biases gradients toward the dominant resolution bucket.
- Cover the text/UI domain explicitly. Reuse the synthetic generator from `coefficient/examples/generate_scale_pyramid` driven on text/UI sources (see `benchmarks/scale_invariance_text-ui_2026-04-17.csv` for the corpus characteristics).

### 2. Feature / weight layer

Two options worth comparing:

- **(a) Add `log₂(pixel_count)` and its squared term as features.** Cheapest. The trainer would learn whether to use them. If they end up with near-zero weight, we lose nothing; if they get nonzero weight, we have an explicit scale correction baked into the score.
- **(b) Per-scale weight normalization.** The current 4-scale × 3-channel × 19-feature layout produces 228 weights that are effectively pooled across scales. Allowing per-scale residual gain terms (one bias per scale, separate from the existing per-feature weights) could let the metric explicitly compensate for "this scale gets noisier slopes." Less interpretable than (a) but possibly more powerful.

Start with (a) — it's a one-line change to feature extraction and forces the question to be answered by the optimizer.

### 3. Objective layer

Add a **scale-stability regularizer** to the training objective:

```
loss = -SROCC(scores, human_ratings)              # existing
       + λ * E_{(source, codec, q)} [ Var_levels( score(level) ) ]
```

where the second term penalizes within-group score variance across pyramid levels for the same (source, codec, quality). λ trades off rank correlation vs scale invariance. We can sweep λ and report a Pareto curve (SROCC vs |β|) so the choice is data-driven.

## Acceptance criteria

- After refit, run `zensim-validate --scale-invariance` on the same matrices used in #12 (photos + text/UI) and report:
  1. Photo-codec median |β| reduced by ≥30% with no SROCC regression on KADIK10K + TID2013 holdouts.
  2. Text/UI-codec median |β| reduced by ≥40% (currently 0.71, target ≤0.40).
  3. SSIMULACRA2 R² collapse on text-codec is a separate metric problem we can't fix here, but zensim's text-codec R² should remain ≥0.7.
- Commit the new weights as a new `WEIGHTS_PREVIEW_V0_3` constant and gate adoption behind a comparison run.

## Related

- #12 — original scale-invariance evaluation
- `docs/scale-invariance.md` — protocol + first-run + follow-up findings
- `benchmarks/scale_invariance_*_2026-04-17.csv` — baseline measurements

## Non-goals

- Fixing SSIMULACRA2 or butteraugli scale bias — those are upstream metrics.
- Adding new feature *channels* (e.g., variance maps, edge density) — separate issue.
- Reducing the existing 228 weights to a smaller set — separate issue.

**Comment by lilith** [comment]
Status check — not started.

Verified absent on main:
- No `log2`-based feature in `zensim/src/metric.rs` feature extraction (grep for `log2.*pixel` / `pixel.*log2` / `log_pixel_count` returns no hits).
- No `WEIGHTS_PREVIEW_V0_3` constant in zensim/src/profile.rs (still on V0_1, 228 weights).
- `zensim-validate` has no scale-stability regularizer.

Acceptance criteria from the issue body still apply unchanged. Prerequisites (resolution-balanced augmentation, text/UI training corpus reuse, training pipeline plumbing) all need to land first.

**Comment by lilith** [comment]
## Implemented and tested in PR #24 (V0_4 v4 branch)

Implemented size axes as 4 appended features for the V0_4 MLP path:
- `log2(pixel_count)`
- `log2(min_dim)`
- `log2(max_dim)`
- signed `log2(max_dim/min_dim)` (aspect, oriented by which dim is wider)

Plumbing:
- `--mlp-size-axes` flag in zensim-validate
- DatasetWithFeatures gains a `dimensions: Vec<(u32, u32)>` field, populated by probing reference image PNG headers (memoized)
- Runtime `apply_mlp_scoring` in zensim::metric detects 232-input MLPs and appends size axes from `source.width()`/`source.height()` before forward
- Trainer-side and runtime-side size-axis encoding are byte-identical (cross-checked)

### Result: marginal-to-negative

V0_4 v3 (no size axes, 228 features) vs V0_4 v4 (with size axes, 232 features) on identical training setup:

| Dataset | v3 | v4 | Δ |
|---------|:---:|:---:|:---:|
| KADIK_val | 0.9350 | 0.8999 | **-0.035** |
| TID_val | 0.9370 | 0.9072 | **-0.030** |
| CID22 | 0.8928 | 0.8933 | +0.0005 |

The 4 size-axis features pushed the optimizer into a different region — a CID22-favorable spot at the cost of KADIK/TID. The min-validation policy correctly picked it (CID22 was the floor metric), but the trade is net-negative on holdout average.

### Why didn't size axes help more?

A few possibilities, untested:
1. **Encoding choice may be wrong.** Pure log2(dim) is invariant under rescaling but doesn't capture viewing-distance semantics. The pixel-angular-size primitive (#25) might be a better signal — same content at different DPRs is perceptually different even at the same pixel count.
2. **Per-corpus scale homogeneity.** KADIK and TID are largely fixed-resolution corpora (KADIK 512×384, TID 512×384). Inside those datasets the size axes are nearly constant — model learns nothing. CID22 has more variation but only 49 refs.
3. **Architecture too small.** With n_hidden=64 and only 228 already-correlated input features, adding 4 more doesn't change the model's effective capacity in a useful way.
4. **Loss function indifferent.** RankNet pairwise loss is rank-only — if a feature shifts every prediction by the same amount, it's invisible to ranking. Per-pair size axes are partially that.

### Where this goes

The infrastructure is in PR #24 and stays. For V0_5 we should try:
- **Pixel-angular-size as an input feature** (requires viewing-condition data per #25's app)
- **Size axes as conditioning** (concatenate to hidden layer rather than input)
- **Multi-scale stratified MLP** — small model per content-class × resolution-band, dispatch at score time

Closing this issue as the immediate "add log2(pixel_count) as input" experiment; the broader scale-bias question stays open for V0_5.
